### ISSUE TITLE
Dotnet-Monitor no longer emitting metrics on port 52325

### ISSUE THREAD

**Issue by webbsk** [dotnet/dotnet-monitor#803]
### Description

DotNet-Monitor no longer seems to emit on 52325 without auth? We needed to switch to another custom port to see them again.

### Regression?

Yes but it's unclear when this broke (within the last couple months)

### Other information

https://microsoft.visualstudio.com/Universal%20Store/_git/SC.CST.ConsumerRecurrenceProcessor/commit/7d58fe46b752b92163a98522a1bbdc5fcf20b704?refName=refs%2Fheads%2Fmaster

**Comment by jander-msft** [comment]
Can you explain "no longer seems to emit"? Is that the URI returns 404 or that it returns an empty response?

**Comment by jander-msft** [comment]
If it is returning an empty response, one possible cause is that the default process was not configured correctly. For example, the `DefaultProcess` configuration entries should not have a `DotnetMonitor_` prefix when used for non-environement-variable configuration.

**Comment by webbsk** [comment]
I end up not being able to connect to the port at all, not an empty response.

That is a good catch with that service having an incorrect prefix, but the other service that I'm looking at does not have those prefixes (and does not have the custom port 52326 port specified) and it is exhibiting the same behaviour: https://dev.azure.com/microsoft/Universal%20Store/_git/SC.Mdollar.Service.BigCatFD?path=%2Fsrc%2FDisplayCatalog.Infrastructure%2Fcharts%2Ffdinfra%2Ftemplates%2Fdotnetmonitorconfigmap.yaml

**Comment by webbsk** [comment]
A little bit more information, if I port forward to the cluster, the results come back (or curl within the same container). However, if I curl using a different container in the same pod or try to connect from a different pod in the same cluster, that's when the failure happens:

![image](https://user-images.githubusercontent.com/1261551/131576684-a9ca95cf-a780-453a-bb7b-14e653c8d827.png)

**Comment by jander-msft** [comment]
> A little bit more information, if I port forward to the cluster, the results come back (or curl within the same container). However, if I curl using a different container in the same pod or try to connect from a different pod in the same cluster, that's when the failure happens. This error is from the dcatfd container in the same pod:
> 
> ![image](https://user-images.githubusercontent.com/1261551/131576684-a9ca95cf-a780-453a-bb7b-14e653c8d827.png)

It looks like you are attempting to `curl` to the target pod via `http://localhost:52323`. When dotnet-monitor is running within a container, its metrics endpoint is not bound to localhost but is bound to the IP address assigned to the container/pod. In a different container within the pod, that might be a distinction without a difference, but that definitely matters when trying to access it from another pod. Also, if accessing from another pod, the port might need to be exposed or possibly exposed via a service (preferably a non-public facing service).

You should also check the logs of the dotnet-monitor container to ensure that it is actually binding the metrics endpoint.

**Comment by webbsk** [comment]
That's right, I was using localhost. Once I've switched to the Cluster-IP (the one that Envoy -our ingress component - is using to bind to the K8S service for dcatfd) 10.0.119.154, I get a similar error:

![image](https://user-images.githubusercontent.com/1261551/131579526-36c7cae2-9cea-4204-a76f-63b446720cab.png)

**Comment by webbsk** [comment]
Here are the logs from the dotnet-monitor container from that pod:
[dotnetmonitorlog.txt](https://github.com/dotnet/dotnet-monitor/files/7086956/dotnetmonitorlog.txt)

**Comment by jander-msft** [comment]
You'll also need to `curl` to the `metrics` route, not just to the host name and port e.g. `curl -v http://10.0.119.154:52325/metrics`

**Comment by webbsk** [comment]
Sorry! That does seem to work then in the same pod, different container. 

However, any ideas why am I seeing all the errors in the logs like this:
{"Timestamp":"2021-08-31T22:03:05.1903487Z","EventId":7,"LogLevel":"Information","Category":"Microsoft.Diagnostics.Tools.Monitor.ApiKeyAuthenticationHandler","Message":"MonitorApiKey was not authenticated. Failure message: API key authentication not configured.","State":{"Message":"MonitorApiKey was not authenticated. Failure message: API key authentication not configured.","AuthenticationScheme":"MonitorApiKey","FailureMessage":"API key authentication not configured.","{OriginalFormat}":"{AuthenticationScheme} was not authenticated. Failure message: {FailureMessage}"},"Scopes":[{"Message":"ConnectionId:0HMBCO6H2O8VQ","ConnectionId":"0HMBCO6H2O8VQ"},{"Message":"RequestPath:/metrics RequestId:0HMBCO6H2O8VQ:00000001, SpanId:|72892409-40495342a454dac0., TraceId:72892409-40495342a454dac0, ParentId:","RequestId":"0HMBCO6H2O8VQ:00000001","RequestPath":"/metrics","SpanId":"|72892409-40495342a454dac0.","TraceId":"72892409-40495342a454dac0","ParentId":""}]}

**Comment by jander-msft** [comment]
That indicates that the `/metrics` route is being accessed via an authenticated URL. All urls that are bound via the `-u` or `--urls` parameter will be restricted to use authentication, even if accessing the `/metrics` route; the default URL is `https://localhost:52323`. All urls that are bound via the `--metricUrls` parameter only allow access to unauthenticated routes, at this time is the `/metrics` route; all other routes will 404; the default URL is `http://localhost:52325` (in non-containers) and `http://+:52325` (in containers).

**Comment by webbsk** [comment]
Ok, sorry for the delay: I've tried redeploying with the different settings and I am convinced now that it is a problem on our side with Envoy blocking the metrics URL because it is an address/port combination that doesn't use host headers or a Kubernetes service. That is, this is NOT a DotNet-Monitor issue. Thanks for the advice about the settings though, Justin - Roberto and I have integrated those changes into our respective service Helm charts.
